### PR TITLE
[WIP] Check the length of PTR record content

### LINKED ISSUE #1895: pdnssec check-zone don't report too long subdomains in ptr
Example:

PTR 10.XX.XX.XX.in-addr.arpa 
value:
diesisteinwirklichsehrlangertestumzusehenobdasmagischelimitvoncasechszigzeichenberuecksichtigtwird.test.de.
$wc says 108 in total .. subdomain is 99

pdnssec says:
Checked 27 records of 'XX.XX.XX.in-addr.arpa', 0 errors, 0 warnings. .... 

but the PTR is not delivered, if I request with dig

gmysql backend with pdns 3.4.0

### PR DESCRIPTION
**Pull request by pieterlexis**
Check for both the total record length and the length of individual
  labels in the content of PTR records. Closes #1895

I don't know if this solution is elegant (not really a C++ guy), but it works:

```
INSERT INTO "domains" VALUES(1,'0.0.10.in-addr.arpa',NULL,NULL,'NATIVE',NULL,NULL);
INSERT INTO "records" VALUES(1,1,'0.0.10.in-addr.arpa','SOA','ns1.plexis.eu hostmaster.plexis.eu 2015010501 86400 7200 604800 300',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(2,1,'0.0.10.in-addr.arpa','NS','ns1.plexis.eu',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(3,1,'1.0.0.10.in-addr.arpa','PTR','123456789012345678901234567890123456789012345678901234567890.plexis.eu',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(4,1,'2.0.0.10.in-addr.arpa','PTR','123456789012345678901234567890.plexis.eu',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(5,1,'3.0.0.10.in-addr.arpa','PTR','1234567890123456789012345678901234567890123456789012345678901234567890.plexis.eu',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(6,1,'4.0.0.10.in-addr.arpa','PTR','12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789.012345678901234567890.12345678901234567890.12345678901234567890.plexis.eu',3600,NULL,NULL,0,NULL,1);
INSERT INTO "records" VALUES(7,1,'5.0.0.10.in-addr.arpa','PTR','12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890',3600,NULL,NULL,0,NULL,1);
```

```
$ ./pdnssec --config-dir /tmp check-zone 0.0.10.in-addr.arpa
[Error] Label in record content too long: 3.0.0.10.in-addr.arpa IN PTR 1234567890123456789012345678901234567890123456789012345678901234567890.plexis.eu
[Error] Total length for PTR record content too long: 4.0.0.10.in-addr.arpa IN PTR 12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789.012345678901234567890.12345678901234567890.12345678901234567890.plexis.eu
[Error] Label in record content too long: 5.0.0.10.in-addr.arpa IN PTR 12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890
Checked 7 records of '0.0.10.in-addr.arpa', 3 errors, 0 warnings.
```

ToDo:
- [ ] Write tests

### CODE REVIEW

**Comment by pieterlexis** [comment]
Hrm, just realized there should be a SERVFAIL sent when these broken records are requested by a client (AKA put this logic into the nameserver itself)

**Comment by Habbie** [comment]
related to #633

**Comment by cmouse** [comment]
Isn't the problem more that label component size exceeds 255, not the total length?

**Comment by cmouse** [comment]
Also, please do add a test in test-dnsrecords_cc.cc

**Comment by Habbie** [comment]
Re SERVFAIL, absolutely - we should not be sending out broken packets :)

**Comment by Habbie** [comment]
In fact if you make dnswriter refuse to do this, check-zone will gain the reporting automatically.

**Comment by pieterlexis** [comment]
Habbie, now I know where to look, I'll see what I can do

**Comment by Habbie** [comment]
I think https://github.com/PowerDNS/pdns/blob/4185c68a927ade027e59c628599faeb404a61a7c/pdns/dnswriter.cc#L256 is the right place, right next to the check to avoid a zero-length label

**Comment by Habbie** [comment]
Also perhaps the `> 255` a few lines earlier should be 63?

**Comment by pieterlexis** [comment]
```
$ ./pdnssec --config-dir /tmp check-zone 0.0.10.in-addr.arpa
[Error] Following record had a problem: 3.0.0.10.in-addr.arpa IN PTR 1234567890123456789012345678901234567890123456789012345678901234567890.plexis.eu
[Error] Error was: DNSPacketWriter::xfrLabel() found a too long label in name
[Error] Following record had a problem: 4.0.0.10.in-addr.arpa IN PTR 12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789.012345678901234567890.12345678901234567890.12345678901234567890.plexis.eu
[Error] Error was: Total length of the name exceeds 255 bytes
[Error] Following record had a problem: 5.0.0.10.in-addr.arpa IN PTR 12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.12345678901234567890.1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890
[Error] Error was: DNSPacketWriter::xfrLabel() found a too long label in name
Checked 7 records of '0.0.10.in-addr.arpa', 3 errors, 0 warnings.
```

```
Jan 06 09:02:19 Exception: DNSPacketWriter::xfrLabel() found a too long label in name
Jan 06 09:02:19 Exception building answer packet (DNSPacketWriter::xfrLabel() found a too long label in name) sending out servfail
Jan 06 09:02:22 Remote ::1 wants '4.0.0.10.in-addr.arpa|PTR', do = 0, bufsize = 1680: packetcache MISS
Jan 06 09:02:22 Exception: Total length of the name exceeds 255 bytes
Jan 06 09:02:22 Exception building answer packet (Total length of the name exceeds 255 bytes) sending out servfail
```



That seems to do the trick, now

File: pdns/dnswriter.cc [hunk 1]
```diff
@@ -236,25 +236,30 @@ void DNSPacketWriter::xfrLabel(const string& Label, bool compress)
       d_labelmap.push_back(make_pair(chopped, pos));                       //  if untrue, we need to count - also, don't store offsets > 16384, won't work
     }
 
+    if(label.size() > 255)
```
Comment (Habbie) [inline pdns/dnswriter.cc]: This ignores that the name may be escaped (see the if below). How about comparing d_record.size before and after?

**Comment by pieterlexis** [comment]
hrm, it looks like the test didn't throw an exception.....
```
Checking bad value for record type PTR test #1

test-dnsrecords_cc.cc(247): error in "test_record_types_bad_values": exception std::runtime_error is expected

Checking bad value for record type PTR test #2

test-dnsrecords_cc.cc(247): error in "test_record_types_bad_values": exception std::runtime_error is expected
```

**Comment by cmouse** [comment]
Bad value test expects that the record parser throws exception. The test itself won't throw anything. If you are throwing same exception as the other DNS record parser/writers then I can assume your code is wrong.

**Pull request closed without merging**